Consolidate template actions components

ntsekouras opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
This PR refactors the code to use the same `TemplateActions` component, that is used in templates and template parts `more` menu(ellipsis). Currently there is a mismatch in provided actions between the sidebar navigation and the template/template part cart menu.


## Notes
In trunk `TemplateActions` are also used for patterns, which is not ideal since it works for template parts but not for the rest of patterns. We need to provide the available actions there too for patterns.

## Testing Instructions
1. Navigate through templates and template parts and observe that the same actions are provided in both places and work as expected.

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/16275880/8e28d59b-2fd6-4f3a-94cc-e9f4cdd23f9d

ntsekouras committed:
Consolidate template actions components

ntsekouras commented:
It seems there was a small difference between the two components. In sidebar navigation when we `cleared the customizations`, we also saved immediately. In the inspector sidebar we wouldn't save and also show the `undo` button in the `snackbar notice`. Is this something we want to preserve or it's fine to revert and save in both cases? --cc @jameskoster  @jasmussen .

jameskoster commented:
Good question. As we look to consolidate the details panel and the Inspector it would be good to align the interaction at some point. I don't necessarily think it has to happen _right now_ but if it's simple enough to implement I'd lean towards treating 'Clear customizations' like a delete action. IE it is instantaneous, but shows a confirm dialog. I think the confirm dialog is particularly important in the editing context, given that most changes are _not_ published until you save.

ntsekouras commented:
>I'd lean towards treating 'Clear customizations' like a delete action. IE it is instantaneous, but shows a confirm dialog.

We have the confirmation dialog. The difference between the two places of the same action(revert) in trunk is:
1. Navigation panel -> saves immediately
2. Inpector controls-> just reverts without saving

In this PR both places save immediately and there is always the options to `undo` of course from the top toolbar.

jameskoster commented:
> In this PR both places save immediately and there is always the options to undo of course from the top toolbar.

That seems okay to me, as long as we display a confirm dialog which alerts the user to the gravity of the action, and offers a way to cancel. 

FWIW I'm _not_ seeing a confirm dialog on this PR, but I am seeing the immediate saving, and undo _doesn't_ seem to work – it just empties the whole template:

https://github.com/WordPress/gutenberg/assets/846565/02611a4f-8da9-4282-a2dc-ba759b6f75fa


I think the thing to avoid is different behaviors in different contexts, given we want to consolidate the Inspector + details panel, and potentially provide access to it from data views for quick editing.

ntsekouras commented:
>FWIW I'm not seeing a confirm dialog on this PR, but I am seeing the immediate saving,

Oh.. I'm sorry about that. I was thinking the `delete` action.. Will look into that.

ntsekouras committed:
use confirm dialog for clearing customizations

ntsekouras commented:
I updated the code to use a confirm dialog when clearing the customizations and I updated the e2e tests.

As a follow up we should see to be more consistent with the `copy` in pattern and page actions too. For example in pages list we have `reset` as label and we should also show a confirm dialog.

jameskoster commented:
I see the confirm dialog now 👍 

There's still an issue where the template is emptied on undo though. To replicate:

1. Customize a theme template + save
2. Refresh the editor
3. Clear customizations
4. Undo
5. See empty template.

> For example in pages list we have reset as label

I think we could probably use 'Reset' instead of 'Clear customizations' for templates, and move the "Use template as supplied by the theme" messaging to the confirm dialog.

ntsekouras commented:
>There's still an issue where the template is emptied on undo though. To replicate:

@jameskoster is the same issue on trunk too? Because I couldn't replicate and we have e2e tests for that too..

jameskoster commented:
Yes I see it on trunk too 🤔 

https://github.com/WordPress/gutenberg/assets/846565/4057eb66-f5fa-4981-9cd7-a8f462722a9f

ntsekouras commented:
>Yes I see it on trunk too 🤔

Hm.. okay. Probably it's better to address this separately then.

>I think we could probably use 'Reset' instead of 'Clear customizations' for templates, and move the "Use template as supplied by the theme" messaging to the confirm dialog. Quick mockup:

We can do that in a follow up, no?

t-hamano commented on packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js:
```diff
@@ -100,7 +104,19 @@ export default function TemplatePanel() {
+							postType={ postType }
+							postId={ postId }
+							className="edit-site-template-card__actions"
+							toggleProps={ { isSmall: true } }
```
```suggestion
							toggleProps={ { size: 'small' } }
```

Although unrelated to this PR, we may want to avoid the prop that is internally deprecated. See #53560

t-hamano commented on test/e2e/specs/site-editor/template-revert.spec.js:
```diff
@@ -37,7 +37,6 @@ test.describe( 'Template Revert', () => {
 		} );
 		await editor.saveSiteEditorEntities();
 		await templateRevertUtils.revertTemplate();
-		await editor.saveSiteEditorEntities();
```
Is there a reason why `saveSiteEditorEntities` was removed?

ntsekouras replied:
>It seems there was a small difference between the two components. In sidebar navigation when we cleared the customizations, we also saved immediately. In the inspector sidebar we wouldn't save and also show the undo button in the snackbar notice

We ended up saving immediately everywhere.

t-hamano commented on test/e2e/specs/site-editor/template-revert.spec.js:
```diff
@@ -129,32 +125,6 @@ test.describe( 'Template Revert', () => {
 		expect( contentAfterUndo ).toEqual( contentBefore );
 	} );
 
-	test( 'should show the edited content after revert and clicking undo in the notice', async ( {
```
Is there a reason why three tests, including this one, were removed?

ntsekouras replied:
There is no more `undo` in the notice and these tests were involving this.

t-hamano reviewed (commented):
I remember a similar approach being attempted when the data view was being implemented (#54271). Now that the data view is stable on all pages, @ramonjd might have some good ideas for this refactoring 🤔

ntsekouras committed:
Update packages/edit-site/src/components/sidebar-edit-mode/template-panel/index.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

jorgefilipecosta commented on packages/edit-site/src/components/template-actions/index.js:
```diff
@@ -82,48 +121,26 @@ export default function TemplateActions( {
-		</DropdownMenu>
+		<>
+			<MenuItem
+				info={ __( 'Use the template as supplied by the theme.' ) }
```
This text is also showing for template parts. Should we have a conditional message and show "Use the template-part as supplied by the theme."?

ntsekouras replied:
Let's do that in a follow up that will update the confirmation modal for this.

jorgefilipecosta reviewed (approved):
Things tested well and the code looks good 👍

ntsekouras commented:
@jameskoster in your suggested design for updating the confirm dialog, it's quite different in terms of design with all the other usages we have in our codebase. 

Although I find the new design better personally, should we add it like this right now for just this use case?

Do we plan on updating the other similar dialog components and/or even consider updating the `@worpdress/components` with the new design?
